### ISSUE TITLE
Pathname() calls to_path, not path

### ISSUE THREAD

**Issue by Nakilon** [googleapis/google-cloud-ruby#11644]
#### Environment details

  - Ruby version: 2.3.8 but it's the same in master
  - Gem name and version: google-cloud-storage v1.21.1 but it's the same in master

#### Steps to reproduce

I'm trying to reuse the gem method `Google::Cloud::Storage::File::Verifier.md5_for` before uploading the file to a bucket instead of reinventing the wheel of applying the Digest library.

#### Code example

```ruby
Google::Cloud::Storage::File::Verifier.md5_for Struct.new(:path).new "/Users/nakilon/.bash_profile"
```
```
TypeError: no implicit conversion of #<Class:0x00007f904663e2e0> into String
```
```ruby
Google::Cloud::Storage::File::Verifier.md5_for Struct.new(:to_path).new "/Users/nakilon/.bash_profile"
```
```
NoMethodError: undefined method `rewind' for #<struct to_path="/Users/nakilon/.bash_profile">
```

The reason of the above error is that by checking for responding to `path` we decide whether to apply `Pathname` but it's wrong because it calls `to_path` so it works only after you pass both:

```ruby
Google::Cloud::Storage::File::Verifier.md5_for Struct.new(:path, :to_path).new "/Users/nakilon/.bash_profile", "/Users/nakilon/.bash_profile"
```
```
=> "UDB/PsrPdibd0i6K9+nR/w=="
```

---

I made a pull quest.

Calling a proper method resolves the issue.
See https://github.com/ruby/ruby/blob/17840baed96a296c0e0678a8c686a228859fa7ac/ext/pathname/pathname.c#L35

```c
path_initialize(VALUE self, VALUE arg) {
...
        str = rb_check_funcall(arg, id_to_path, 0, NULL);
```

**Comment by quartzmo** [comment]
@Nakilon Thanks for opening this issue and the accompanying PR.

`Google::cloud::Storage::File::Verifier` is currently an undocumented internal class, so the first thing to look at is whether it should be documented and made public for the use case of verifying file data before upload.

**Comment by Nakilon** [comment]
BTW, after this patch passing the `Pathname` starts working too:
```ruby 
Google::Cloud::Storage::File::Verifier.md5_for Pathname.new "/Users/nakilon/.bash_profile"
```
```
=> "UDB/PsrPdibd0i6K9+nR/w=="
```
otherwise it threw the same error:
```
NoMethodError: undefined method `rewind' for #<Pathname:/Users/nakilon/.bash_profile>
```

**Comment by quartzmo** [comment]
@frankyn Do any of the other client libraries offer verification helpers? How do you feel about adding this to the client surface?

**Comment by frankyn** [comment]
I think this is a gap in the Ruby client, but the create_file method should expose a flag to handle md5/crc32c and verify integrity after upload is completed. I wouldn't expose the helper outside of an additional parameter to do checksum validation on uploads.

**Comment by quartzmo** [comment]
> I think this is a gap in the Ruby client, but the create_file method should expose a flag to handle md5/crc32c and verify integrity after upload is completed. 

@frankyn The Ruby client has [`crc32c` and `md5` options](https://github.com/googleapis/google-cloud-ruby/blob/master/google-cloud-storage/lib/google/cloud/storage/bucket.rb#L1146-L1157) to accept externally calculated values in `Bucket#create_file`. Is that what you mean?

**Comment by frankyn** [comment]
It would be to provide support to calculate these values on behalf of the user instead of having the user compute them beforehand.

**Comment by Nakilon** [comment]
I'll just remind that the bug and the question of your lib/OOP design are different things. No matter if the method private or public the if condition checks for wrong method respondence as far as I understand.

**Comment by quartzmo** [comment]
@Nakilon I agree, I added a new issue (link above) for @frankyn's feature request so that this issue can remain focused on your issue.
